decode.go

```diff
@@ -123,166 +135,234 @@ func parseDateTime(b []byte) (time.Time, error) {
 	}
 
 	tok := string(b)
-	err := numberContainsInvalidUnderscore(tok)
-	if err != nil {
+	if err := numberContainsInvalidUnderscore(tok); err != nil {
 		return 0, err
 	}
+
 	cleanedVal := cleanupNumberToken(tok)
 	if cleanedVal[0] == '.' {
-		return 0, fmt.Errorf("float cannot start with a dot")
+		return 0, ErrParseFloatStartDot
 	}
+
 	if cleanedVal[len(cleanedVal)-1] == '.' {
-		return 0, fmt.Errorf("float cannot end with a dot")
+		return 0, ErrParseFloatEndDot
+	}
+
+	f, err := strconv.ParseFloat(cleanedVal, 64)
+	if err != nil {
+		return 0, fmt.Errorf("coudn't ParseFloat %w", err)
 	}
-	return strconv.ParseFloat(cleanedVal, 64)
+
+	return f, nil
 }
 
 func parseIntHex(b []byte) (int64, error) {
 	tok := string(b)
 	cleanedVal := cleanupNumberToken(tok)
-	err := hexNumberContainsInvalidUnderscore(cleanedVal)
+
+	if err := hexNumberContainsInvalidUnderscore(cleanedVal); err != nil {
+		return 0, err
+	}
+
+	i, err := strconv.ParseInt(cleanedVal[2:], 16, 64)
 	if err != nil {
-		return 0, nil
+		return 0, fmt.Errorf("coudn't ParseIntHex %w", err)
 	}
-	return strconv.ParseInt(cleanedVal[2:], 16, 64)
+
+	return i, nil
 }
 
 func parseIntOct(b []byte) (int64, error) {
 	tok := string(b)
 	cleanedVal := cleanupNumberToken(tok)
-	err := numberContainsInvalidUnderscore(cleanedVal)
-	if err != nil {
+
+	if err := numberContainsInvalidUnderscore(cleanedVal); err != nil {
 		return 0, err
 	}
-	return strconv.ParseInt(cleanedVal[2:], 8, 64)
+
+	i, err := strconv.ParseInt(cleanedVal[2:], 8, 64)
+	if err != nil {
+		return 0, fmt.Errorf("coudn't ParseIntOct %w", err)
+	}
+
+	return i, nil
 }
 
 func parseIntBin(b []byte) (int64, error) {
 	tok := string(b)
 	cleanedVal := cleanupNumberToken(tok)
-	err := numberContainsInvalidUnderscore(cleanedVal)
-	if err != nil {
+
+	if err := numberContainsInvalidUnderscore(cleanedVal); err != nil {
 		return 0, err
 	}
-	return strconv.ParseInt(cleanedVal[2:], 2, 64)
+
+	i, err := strconv.ParseInt(cleanedVal[2:], 2, 64)
+	if err != nil {
+		return 0, fmt.Errorf("coudn't ParseIntBin %w", err)
+	}
+
+	return i, nil
 }
 
 func parseIntDec(b []byte) (int64, error) {
 	tok := string(b)
 	cleanedVal := cleanupNumberToken(tok)
-	err := numberContainsInvalidUnderscore(cleanedVal)
-	if err != nil {
+
+	if err := numberContainsInvalidUnderscore(cleanedVal); err != nil {
 		return 0, err
 	}
-	return strconv.ParseInt(cleanedVal, 10, 64)
+
+	i, err := strconv.ParseInt(cleanedVal, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("coudn't parseIntDec %w", err)
+	}
+
+	return i, nil
 }
 
 func numberContainsInvalidUnderscore(value string) error {
 	// For large numbers, you may use underscores between digits to enhance
 	// readability. Each underscore must be surrounded by at least one digit on
 	// each side.
-
 	hasBefore := false
+
 	for idx, r := range value {
 		if r == '_' {
 			if !hasBefore || idx+1 >= len(value) {
 				// can't end with an underscore
 				return errInvalidUnderscore
 			}
 		}
+
```

Comment:
Out of curiosity, are those newlines enforced by golangci-lint or is this personal preference?

Reply:
golangci-lint, I won't enforce any personal opinions :).